### PR TITLE
Switch to use the new SSP operator 

### PR DESCRIPTION
**Pull request by zcahana**
This PR updates HCO operator to use the new [Golang-based SSP operator](https://github.com/kubevirt/ssp-operator), instead of the old [Ansible-based SSP operator](https://github.com/kubevirt/kubevirt-ssp-operator).

~Note 1: This isn't mergeable yet, since the new SSP operator doesn't have a released image yet.~
~Note 2: I plan to add additional unit tests to cover the upgrade logic (removal of old CRDs, etc).~

```release-note
Switch to use the new SSP operator.
```

### CODE REVIEW

File: pkg/controller/operands/ssp.go
Comment (zcahana) [inline pkg/controller/operands/ssp.go:3]: Github's diff view of this file is terrible.
In essence, this file has been rewritten from scratch, so it better be viewed as a whole file, and compared to the old file as a whole.

File: cmd/hyperconverged-cluster-operator/main.go
Comment (tiraboschi) [inline cmd/hyperconverged-cluster-operator/main.go:34]: I fear we can accept it only if >= v1beta1

Comment (zcahana) [reply cmd/hyperconverged-cluster-operator/main.go:34]: What is the criteria to move from v1alpha1 to v1beta1? Or this is merely a syntactical requirement?

**Review by nunnatsa — COMMENTED** [review]
Starting the review. Currently - two comments.

File: pkg/controller/operands/ssp.go
Comment (zcahana) [reply pkg/controller/operands/ssp.go:66]: I too don't really like that global var and the appends. These are basically Go's poorman's constant slice and slice copy.
As you said this isn't much of a big deal, so I'll just inline the CRD names...

File: pkg/controller/operands/ssp.go
Comment (zcahana) [reply pkg/controller/operands/ssp.go:78]: Yes this cries out for concurrency. I think this is simple enough so better be implemented now rather than postponed to sometimes between [unknown, never].

File: cmd/hyperconverged-cluster-operator/main.go
Comment (tiraboschi) [reply cmd/hyperconverged-cluster-operator/main.go:34]: The criteria is:

Since it is a "beta" suffixed, it should work with subsequent releases. Indeed, the meaning of api versions is as following:

alpha
API versions with ‘alpha’ in their name are early candidates for new functionality coming into Kubernetes. These may contain bugs and are not guaranteed to work in the future.

beta
‘beta’ in the API version name means that testing has progressed past alpha level, and that the feature will eventually be included in Kubernetes. Although the way it works might change, and the way objects are defined may change completely, the feature itself is highly likely to make it into Kubernetes in some form.

stable
These do not contain ‘alpha’ or ‘beta’ in their name. They are safe to use.

an operator is required to have it's CRD at least to v1beta1 to be considered GA as our previous version was.

File: pkg/controller/operands/ssp.go
Comment (zcahana) [reply pkg/controller/operands/ssp.go]: Done.

File: pkg/controller/operands/ssp.go
Comment (zcahana) [reply pkg/controller/operands/ssp.go]: Done.

File: pkg/controller/operands/ssp.go
Comment (zcahana) [reply pkg/controller/operands/ssp.go]: By the way, the way I implemented this assumes that the controller-runtime's Client is thread-safe. I'm not sure whether this assumption is correct or not, but [asked](https://groups.google.com/u/3/g/kubebuilder/c/CT0wdZykPwc).

**Comment by zcahana** [comment]
Rebased + several updates.

File: pkg/controller/operands/ssp.go
Comment (zcahana) [reply pkg/controller/operands/ssp.go]: Here's the answer I got from one of the controller-runtime devs (for some reason it doesn't appear on the google group thread):

_Yes, the delegating client is safe to use concurrently. The delegating client itself doesn't have any moving parts after its creation, and the underlying clients (cached, and live) should be already safe._

File: cmd/hyperconverged-cluster-operator/main.go
Comment (akrejcir) [reply cmd/hyperconverged-cluster-operator/main.go:34]: The API version of `ssp-operator` is now `v1beta1`.

Comment (zcahana) [reply cmd/hyperconverged-cluster-operator/main.go:34]: @akrejcir thanks, will update this PR to consume `v1beta1`.

**Comment by zcahana** [comment]
/hold

We need to revert efba5d2 before merging this PR - and only after the fix is applied and released in [kubevirt/ssp-operator](https://github.com/kubevirt/ssp-operator).

**Comment by zcahana** [comment]
@tiraboschi per our discussion, should we go ahead an /override the two upgrade tests?

File: pkg/controller/operands/ssp_test.go
Comment (nunnatsa) [inline pkg/controller/operands/ssp_test.go:113]: This is a bit misleading to me. This Context and the one in line 248 uses the old names and I find it confusing. But I undersand the logic. Your call...

Also, since workloads and infra are not depeneded in each other, do we reaally need this split in the contexts. Maybe we do it inside the same one?

File: pkg/controller/operands/ssp_test.go
Comment (nunnatsa) [inline pkg/controller/operands/ssp_test.go:383]: Maybe it worth to add a test case(s) where the SSP CR already exist (for future upgrades). Although this is a very unlikely to be relevant.

File: pkg/controller/operands/ssp_test.go
Comment (zcahana) [reply pkg/controller/operands/ssp_test.go:113]: The use of `NodeLabeller` and `TemplateValidator` don't refer to the old CRDs, but rather to the components (which still exist) and their corresponding fields within `SSP.Spec`.

About merging both under the same context... I was considering this too while implementing this, but thought that it would be good to test for them being and staying independent when the code change. But rethinking this now, I think it'd be better to shrink the test code in ~140 lines vs. future-proofing from unlikely bugs. So, will merge the contexts.

Comment (zcahana) [reply pkg/controller/operands/ssp_test.go:113]: Done.

File: pkg/controller/operands/ssp_test.go
Comment (zcahana) [reply pkg/controller/operands/ssp_test.go:383]: Not sure what is the test case to be added? Meaning, what logic to we want to test?

Comment (nunnatsa) [reply pkg/controller/operands/ssp_test.go:277]: Just make sure that the HyperConverged CR 's node placement related data is copied to the SSP CR

Comment (zcahana) [reply pkg/controller/operands/ssp_test.go:277]: This behavior is already covered in various tests under the "Node placement" context.

**Comment by zcahana** [comment]
Waiting for merge of #1011 so that I can rebase on top of that (instead of cherry-picking its commit into this PR).

**Comment by zcahana** [comment]
@nunnatsa no need to rerun the `upgrade-` tests, these are expected to fail here and it was concluded to /override them.

**Comment by zcahana** [comment]
/test hco-e2e-upgrade-prev-azure 
/test hco-e2e-image-index-azure

flakytest.

**Comment by nunnatsa** [comment]
> @nunnatsa no need to rerun the `upgrade-` tests, these are expected to fail here and it was concluded to /override them.

Right. What about prev ? they must pass. But the same is for all azure test that currently not working

**Comment by zcahana** [comment]
> Right. What about prev ? they must pass, I guess. But the same is for all azure test that currently not working

Yes, prev must pass. All failures seems to be due to CI flakyness.

**Comment by nunnatsa** [comment]
test hco-e2e-upgrade-aws
/test hco-e2e-upgrade-azure

**Comment by zcahana** [comment]
@tiraboschi / @nunnatsa can we already `/override` the upgrade tests, so that we don't accidentally `/retest` their failures?

**Comment by nunnatsa** [comment]
> /test hco-e2e-upgrade-aws

This will fail for sure. It can't pas until we'll update the image in quay.io after merging this one.

File: hack/check_old_ssp_removed.sh
Comment (zcahana) [reply hack/check_old_ssp_removed.sh:60]: This block shouldn't be commented - just a remainder from my local testing of it.
Thanks for spotting this!

File: pkg/controller/commonTestUtils/testUtils.go
Comment (zcahana) [reply pkg/controller/commonTestUtils/testUtils.go]: The dup's purpose is to create `NodePlacement` instance which is different than the one created by the already existing `NewHyperConvergedConfig()` function. The naming is not great, I agree. I wanted to (re)name them `NewNodePlacement()` and `NewOtherNodePlacement()`, but then preferred to avoid touching lots of test files that are already referencing `NewHyperConvergedConfig()` (as this PR is already quite too large). I could use `NewOtherHyperConvergedConfig()`, which is also not great since this isn't creating a HyperConvergedConfig, but at least is still consistent with the original function.

Comment (zcahana) [reply pkg/controller/commonTestUtils/testUtils.go]: Eventually renamed both to `NewNodePlacement()` and `NewOtherNodePlacement()` - took the toll of touching 7-8 additional "unrelated" files.

File: hack/check_old_ssp_removed.sh
Comment (zcahana) [reply hack/check_old_ssp_removed.sh]: Fixed.

**Review by nunnatsa — APPROVED** [review]
/approve

**Comment by nunnatsa** [comment]
/override ci/prow/hco-e2e-upgrade-aws
/override ci/prow/hco-e2e-upgrade-azure

Failure is expected

/override ci/prow/hco-e2e-image-index-gcp
hco-e2e-image-index-azure hco-e2e-image-index-aws passed

**Review by nunnatsa — APPROVED** [review]
/approve

**Pull request merged**
